### PR TITLE
fix: measure ReplicatedMap operations latency in nanoseconds and report in milliseconds

### PR DESCRIPTION
**Pull request by alex-dukhno**
> `LocalReplicatedMapStatsImpl` states that `maxGetLatency` is measured in `MS` but in `ReplicatedMapGetMessageTask` we call `incrementGets` with nanoseconds diff.

As suggested in https://github.com/hazelcast/hazelcast/pull/16795#discussion_r397300346 the PR changes logic how IMDG compute `max*Latency` metrics.

After merging the PR `max*Latency` values are collected in nanoseconds and reported in milliseconds

### CODE REVIEW

**Comment by puzpuzpuz** [comment]
@blazember could you take a look at this PR?

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/replicatedmap/ReplicatedMapGetMessageTask.java [hunk 1]
```diff
@@ -59,7 +60,7 @@ protected Object processResponseBeforeSending(Object response) {
         ReplicatedMapService replicatedMapService = getService(ReplicatedMapService.SERVICE_NAME);
         if (replicatedMapService.getReplicatedMapConfig(parameters.name).isStatisticsEnabled()) {
             LocalReplicatedMapStatsImpl stats = replicatedMapService.getLocalReplicatedMapStatsImpl(parameters.name);
-            stats.incrementGets(System.nanoTime() - startTimeNanos);
+            stats.incrementGets(TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTimeNanos));
```
Comment (blazember) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/replicatedmap/ReplicatedMapGetMessageTask.java]: I think we should rather report it as nanos and change the unit to `NS` too, similarly to the map counterpart: https://github.com/hazelcast/hazelcast/blob/97ff57ba92dc1fbfa77720885e7db40757999080/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapGetMessageTask.java#L71 

I'd expect with reporting as millis, we typically report 0.

The same fields in the map are not metrics though :thinking: This seems to be an error, I send a PR to fix this.
https://github.com/hazelcast/hazelcast/blob/296aac5c8a3a4970851afc55c9e91622db75c131/hazelcast/src/main/java/com/hazelcast/internal/monitor/impl/LocalMapStatsImpl.java#L140

Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/replicatedmap/ReplicatedMapGetMessageTask.java]: After taking another look, we do have metrics for maps' total get latency, but we do it a bit different:
https://github.com/hazelcast/hazelcast/blob/296aac5c8a3a4970851afc55c9e91622db75c131/hazelcast/src/main/java/com/hazelcast/internal/monitor/impl/LocalMapStatsImpl.java#L311-L315

So we accumulate nanos, but convert it to millis for the metrics collector. I believe this is the right way to do.
@alex-dukhno @puzpuzpuz :point_up:

Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/replicatedmap/ReplicatedMapGetMessageTask.java]: > I believe this is the right way to do.

I agree. We should do the same here.

Comment (alex-dukhno) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/replicatedmap/ReplicatedMapGetMessageTask.java]: @blazember how about `MAX_*_LATENCY`? they are collected in the same methods as `total*Latencies`
e.g. https://github.com/hazelcast/hazelcast/blob/296aac5c8a3a4970851afc55c9e91622db75c131/hazelcast/src/main/java/com/hazelcast/internal/monitor/impl/LocalReplicatedMapStatsImpl.java#L238-L242

File: hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/AbstractReplicatedRecordStore.java [hunk 2]
```diff
@@ -75,7 +75,7 @@ public Object removeWithVersion(Object key, long version) {
     @SuppressWarnings("unchecked")
     private Object remove(InternalReplicatedMapStorage<K, V> storage, Object key) {
         isNotNull(key, "key");
-        long time = Clock.currentTimeMillis();
+        long time = System.nanoTime();
```
Comment (puzpuzpuz) [inline hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/AbstractReplicatedRecordStore.java]: Changes in this class seem to be unrelated. #16781 is meant to address the problem across the whole code base.

File: hazelcast/src/main/java/com/hazelcast/internal/monitor/impl/LocalReplicatedMapStatsImpl.java [hunk 3]
```diff
@@ -400,14 +405,22 @@ public void fromJson(JsonObject json) {
...
-        totalRemoveLatencies = getLong(json, "totalRemoveLatencies", -1L);
+        totalGetLatenciesNanos = convertMillisToNanos(getLong(json, "totalGetLatencies", -1L));
+        totalPutLatenciesNanos = convertMillisToNanos(getLong(json, "totalPutLatencies", -1L));
+        totalRemoveLatenciesNanos = convertMillisToNanos(getLong(json, "totalRemoveLatencies", -1L));
         maxGetLatency = getLong(json, "maxGetLatency", -1L);
         maxPutLatency = getLong(json, "maxPutLatency", -1L);
         maxRemoveLatency = getLong(json, "maxRemoveLatency", -1L);
     }
 
+    private static long convertNanosToMillis(long nanos) {
```
Comment (puzpuzpuz) [inline hazelcast/src/main/java/com/hazelcast/internal/monitor/impl/LocalReplicatedMapStatsImpl.java]: nit: extract this and the following one methods into a common module (say, `TimeUtil`) and reuse them here and in `LocalMapStats`.

File: hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/AbstractReplicatedRecordStore.java [hunk 4]
```diff
@@ -86,7 +86,7 @@ private Object remove(InternalReplicatedMapStorage<K, V> storage, Object key) {
             storage.remove(marshalledKey, current);
         }
         if (replicatedMapConfig.isStatisticsEnabled()) {
-            getStats().incrementRemoves(Clock.currentTimeMillis() - time);
+            getStats().incrementRemoves(System.nanoTime() - time);
```
Comment (puzpuzpuz) [inline hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/AbstractReplicatedRecordStore.java]: This call with nanos instead of millis as the argument automatically changes unit for `maxRemoveLatency` (and other `max*Latency` fields) in `LocalReplicatedMapStatsImpl` to nanos. Thus, they should be converted when reading the probe and when serializing stats to JSON, like it's done in `LocalMapStatsImpl`.

Comment (puzpuzpuz) [inline hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/AbstractReplicatedRecordStore.java]: nit: rename argument of `#incrementRemoves()` to `latencyNanos` and, optionally, rename the method itself to `#incrementRemovesLatencyNanos` (see `LocalMapStatsImpl` for example). The same consideration applies to other `#increment*()` methods.

File: hazelcast/src/main/java/com/hazelcast/internal/monitor/impl/LocalReplicatedMapStatsImpl.java [hunk 5]
```diff
@@ -235,62 +232,68 @@ public long getPutOperationCount() {
...
     public long getTotalRemoveLatency() {
-        return totalRemoveLatencies;
+        return convertNanosToMillis(totalRemoveLatenciesNanos);
     }
 
+    @Probe(name = REPLICATED_MAP_MAX_GET_LATENCY, unit = MS)
     @Override
     public long getMaxPutLatency() {
-        return maxPutLatency;
+        return maxPutLatencyNanos;
```
Comment (blazember) [inline hazelcast/src/main/java/com/hazelcast/internal/monitor/impl/LocalReplicatedMapStatsImpl.java]: These should be converted to `MS` too.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/replicatedmap/ReplicatedMapGetMessageTask.java (hunk 1 above)
Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/replicatedmap/ReplicatedMapGetMessageTask.java]: I think `MS` is good, the max is very likely to be in the `MS` range anyways. Also, the map stats report it as `MS`, the best is to be consistent across the structures.

File: hazelcast/src/main/java/com/hazelcast/internal/monitor/impl/LocalReplicatedMapStatsImpl.java (hunk 5 above)
Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/internal/monitor/impl/LocalReplicatedMapStatsImpl.java]: Nice catch. I don't know why I didn't notice this. 😅

**Comment by blazember** [comment]
@alex-dukhno The commit addressing my comment looks good to me, but there are test failures in the build. Could you take a look?

**Comment by alex-dukhno** [comment]
run-lab-run
